add support for blob tx

blockchaindevsh opened the pull request:
Summary of this PR:
1. allow l2 blob tx if `enablel2blob` is specified.
2. add blobs field to `RollupCostData`, and update `newL1CostFuncEcotone` accordingly, basically a blob tx will have two extra fees: da fee and da proof fee.
3. fix `Transaction.RollupCostData` so that the result is consistent when called from worker and state processor.
4. port changes from upstream geth to support estimate gas for blob tx.

zhiqiangxu committed:
enable blob tx

qizhou commented on core/txpool/validation.go:
```diff
@@ -69,7 +69,7 @@ func ValidateTransaction(tx *types.Transaction, head *types.Header, signer types
 		return core.ErrTxTypeNotSupported
 	}
-	if opts.Config.IsOptimism() && tx.Type() == types.BlobTxType {
+	if false && opts.Config.IsOptimism() && tx.Type() == types.BlobTxType {
```
Can be put a flag to enable L2 blob rather than false/true (blow)?

blockchaindevsh replied:
Done, the flag is named `enablel2blob`.

zhiqiangxu committed:
configure 4844 with "enable4844" option

zhiqiangxu committed:
always overwrite

zhiqiangxu committed:
rename flag to "enablel2blob"

zhiqiangxu committed:
add enablel2blob to nodeFlags

zhiqiangxu committed:
support estimate gas for blob tx

zhiqiangxu committed:
1. fix RollupCostData when called from worker.commitBlobTransaction, otherwise the result will be different when called from state processor
  2. l2 blob tx will have two extra fees for L1CostFunc: da fee and da proof fee
  3. only check versionedHashes when it's not nil

qizhou commented on core/types/transaction.go:
```diff
@@ -378,7 +378,7 @@ func (tx *Transaction) RollupCostData() RollupCostData {
 		return v.(RollupCostData)
 	}
-	data, err := tx.MarshalBinary()
+	data, err := tx.WithoutBlobTxSidecar().MarshalBinary()
```
I think we should add a few comments about why use non BLOB (but with BLOB hash) version of tx.

Another question is in which case the tx will contain the BLOB (as the derivation stage, we cannot derive the BLOB from L1).

blockchaindevsh replied:
I've updated with some comments.

qizhou commented on eth/ethconfig/config.go:
```diff
@@ -167,6 +167,8 @@ type Config struct {
 
 	OverrideOptimismInterop *uint64 `toml:",omitempty"`
 
+	EnableL2Blob bool
```
I would add a few comments of like "Enable accepting EIP-4844 BLOBs on L2"

blockchaindevsh replied:
Done.

qizhou commented on params/protocol_params.go:
```diff
@@ -47,6 +47,9 @@ const (
 	LogDataGas            uint64 = 8     // Per byte in a LOG* operation's data.
 	CallStipend           uint64 = 2300  // Free gas given at beginning of call.
 
+	BlobDAFee      = 2000
```
Comments are needed here.

blockchaindevsh replied:
Done.

qizhou commented on ethclient/ethclient.go:
```diff
@@ -664,6 +664,15 @@ func toCallArg(msg ethereum.CallMsg) interface{} {
+	if msg.AccessList != nil {
+		arg["accessList"] = msg.AccessList
+	}
+	if msg.BlobGasFeeCap != nil {
```
What is the purpose of this parameter?

blockchaindevsh replied:
It's used [here](https://github.com/ethereum-optimism/op-geth/blob/84b52dfd2626f5e19da6baa2691d81f4c2299eb5/core/state_transition.go#L378), so that a blob tx will not be included if the specified blob gas fee cap is smaller than `BlobBaseFee` of current block.

qzhodl replied:
Geth v1.13.13 supported [it](https://github.com/ethereum/go-ethereum/pull/28989/files) on Feb 21, and op-geth has not merged it...

blockchaindevsh replied:
It seems I mis-understood Qi's question.

Here's more context about this change:https://github.com/ethereum-optimism/op-geth/pull/299

qizhou reviewed (commented):
I feel the description of the PR should be updated.

zhiqiangxu committed:
add more comment

blockchaindevsh commented:
> I feel the description of the PR should be updated.

Done.

qizhou commented on core/types/rollup_cost.go:
```diff
@@ -213,6 +214,7 @@ func newL1CostFuncEcotone(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseF
 		fee = new(big.Int).Add(calldataCostPerByte, blobCostPerByte)
 		fee = fee.Mul(fee, calldataGasUsed)
 		fee = fee.Div(fee, ecotoneDivisor)
+		fee = fee.Add(fee, big.NewInt(int64(costData.blobs*params.BlobDAFee)))
```
Need to deal with integer overflow case?

blockchaindevsh replied:
Good catch, fixed!

qizhou commented on params/protocol_params.go:
```diff
@@ -47,6 +47,9 @@ const (
 	CallStipend           uint64 = 2300  // Free gas given at beginning of call.
 
+	BlobDAFee      = 2000 // Fee for storing blob to DA provider
+	BlobDAProofGas = 200  // Gas for storing DA proof to span batch
```
I would add a comment `It should be average proof size * TxDataNonZeroGasEIP2028`

blockchaindevsh replied:
Added.

qzhodl commented on cmd/geth/config.go:
```diff
@@ -194,6 +194,8 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
 		cfg.Eth.OverrideVerkle = &v
 	}
 
+	cfg.Eth.EnableL2Blob = ctx.Bool(utils.EnableL2Blob.Name)
```
Do we need to check ctx.IsSet()?

blockchaindevsh replied:
I think their effects are the same here.

zhiqiangxu committed:
fix for review

zhiqiangxu committed:
skip check versionedHashes when it's empty

zhiqiangxu committed:
fix "blob transaction without blobs in miner"

blockchaindevsh committed:
remove BlobDAProofGas
